### ISSUE TITLE
Random crashes

### ISSUE THREAD

**Issue by GoogleCodeExporter** [KGIII/pino-twitter#245]
```
What steps will reproduce the problem?
1. Open pino
2. Click on anything
3. Will crash ramdomly clicking on status icon, menu elements or any icon
from the toolbar

What version of the product are you using? On what operating system?

Fedora 12 (ix86)

Please provide any additional information below.

This happens since pino 0.2.6

Starting program: /usr/bin/pino 
[Thread debugging using libthread_db enabled]
[New Thread 0xb7d70b70 (LWP 27291)]
[New Thread 0xb3e65b70 (LWP 27292)]
[New Thread 0xb32ffb70 (LWP 27293)]
[Thread 0xb3e65b70 (LWP 27292) exited]
[Thread 0xb32ffb70 (LWP 27293) exited]
[New Thread 0xb32ffb70 (LWP 27294)]
[New Thread 0xb3e65b70 (LWP 27295)]
[New Thread 0xb26ffb70 (LWP 27296)]
[New Thread 0xb1cfeb70 (LWP 27297)]
[New Thread 0xb12fdb70 (LWP 27298)]
[New Thread 0xb08fcb70 (LWP 27299)]
[New Thread 0xafefbb70 (LWP 27300)]
[New Thread 0xaf4fab70 (LWP 27301)]
[New Thread 0xaeaf9b70 (LWP 27302)]
[New Thread 0xae0f8b70 (LWP 27303)]
[New Thread 0xad6f7b70 (LWP 27304)]
[Thread 0xae0f8b70 (LWP 27303) exited]
[Thread 0xb12fdb70 (LWP 27298) exited]
[Thread 0xb08fcb70 (LWP 27299) exited]
[Thread 0xaeaf9b70 (LWP 27302) exited]
[Thread 0xafefbb70 (LWP 27300) exited]
[Thread 0xb1cfeb70 (LWP 27297) exited]
[Thread 0xb32ffb70 (LWP 27294) exited]
[Thread 0xb3e65b70 (LWP 27295) exited]
[Thread 0xb26ffb70 (LWP 27296) exited]
[Thread 0xaf4fab70 (LWP 27301) exited]
[New Thread 0xaf4fab70 (LWP 27305)]
[New Thread 0xb26ffb70 (LWP 27306)]
[New Thread 0xb3e65b70 (LWP 27307)]
[New Thread 0xb1cfeb70 (LWP 27308)]
[Thread 0xb3e65b70 (LWP 27307) exited]
[Thread 0xb26ffb70 (LWP 27306) exited]
[Thread 0xad6f7b70 (LWP 27304) exited]
Assertion 'pthread_setspecific(t->key, userdata) == 0' failed at
pulsecore/thread-posix.c:196, function pa_tls_set(). Aborting.


```

Original issue reported on code.google.com by `darksh...@gmail.com` on 5 May 2010 at 3:19

**Comment by GoogleCodeExporter** [comment]
```
I'm having the same problem on Ubuntu 10.04 32bit
```

Original comment by `jeffmhub...@gmail.com` on 10 May 2010 at 1:16

**Comment by GoogleCodeExporter** [comment]
```
Issue has become more frequent in the last days, making pino almost unusable :'(

Every time it crashes, this information is shown:

Assertion 'pthread_setspecific(t->key, userdata) == 0' failed at
pulsecore/thread-posix.c:196, function pa_tls_set(). Aborting.

```

Original comment by `darksh...@gmail.com` on 11 May 2010 at 7:51

**Comment by GoogleCodeExporter** [comment]
```
I have done some research. Seems to be related to a bug in pulseaudio and some 
kind
of memory corruption. Reading similar issues on other projects, pa_tls_set() 
function
should not fail unless the key is invalid.

http://www.google.com.mx/search?&q=failed+at+pulsecore%2Fthread-posix.c+function
+pa_tls_set%28%29+Aborting&aq=f&aqi=&aql=&oq=&gs_rfai=

Using gstreamer-properties and setting ALSA as default output plugin, *seems* 
*to*
*solve* *the* *issue*.
```

Original comment by `darksh...@gmail.com` on 11 May 2010 at 8:12

**Comment by GoogleCodeExporter** [comment]
```
+1. 

Ubuntu Lucid 10.04 32-bit, noticing Pino crashes over the last two days. 
Googling for 
the error message printed to stdout brought me here.

I ran gstreamer-properties and changed output to ALSA, but pino is still 
crashing :-( 
Is this ANOTHER PulseAudio problem in Ubuntu? :-(

```

Original comment by `edgar.b....@gmail.com` on 15 May 2010 at 8:49

**Comment by GoogleCodeExporter** [comment]
```
It worked for me (Fedora 12) for a couple of days. Right now it crashes again. 
Seems 
it's a pulseaudio issue, not pino at all.
```

Original comment by `darksh...@gmail.com` on 15 May 2010 at 2:28

**Comment by GoogleCodeExporter** [comment]
```
Pino 0.2.9 still has the issue, still related to pulseaudio, but this time 
there is
something new in the output, related to the key for userdata.

** (pino:12963): WARNING **: Invalid GValue unboxing (wrong type or NULL)

** (pino:12963): WARNING **: Invalid GValue unboxing (wrong type or NULL)
Assertion 'pthread_setspecific(t->key, userdata) == 0' failed at
pulsecore/thread-posix.c:196, function pa_tls_set(). Aborting.

```

Original comment by `darksh...@gmail.com` on 15 May 2010 at 5:31

**Comment by GoogleCodeExporter** [comment]
```
Sorry, but I don't have any idea about this problem. Pino works stable for me 
since 
0.2.7 version.
```

Original comment by `tro...@gmail.com` on 15 May 2010 at 5:43

**Comment by GoogleCodeExporter** [comment]
```
Apparently, seems to be related to the Pulseaudio version used in Fedora 12 and
Ubuntu 10.04.
```

Original comment by `darksh...@gmail.com` on 15 May 2010 at 7:42

**Comment by GoogleCodeExporter** [comment]
```
I added ppa:troorl/pino to software sources for Lucid, updated, and installed 
0.2.9-
ubuntu2~lucid. As you say, darkshram, still the same problem.

I suspected that the new message notifications might be involving PA (from 
libnotify..??) so 
decided to turn off system sounds entirely (Sound Preferences, Sound Effects 
tab, Sound theme: 
"No sounds"). Launched Pino from terminal. Worked for several minutes, through 
refreshes of 
timeline, Prefs dialog and more. I tried turning off all notifications 
checkboxes in Prefs, and 
tried relaunching Pino with Sound Effects turned back on - nope, crashes very 
soon with same old 
PA error. :-(

However, if I start Pino with Sound Effects off, I can then turn Sound Effects 
back on while 
Pino is running, and Pino keeps running (tried this only once, will try again, 
others, could you 
pls see if this works for you?). In any case, I'm not addicted to the Sound 
Effects anyway. I 
couldn't care less if they are turned off; but I want Pino to stop crashing! :-)

The last time around, PA blamed Ubuntu's packaging for the horrible problems we 
had with sound 
in Karmic. This time, this bug is across distros... wonder who's going to be 
blamed... Fedora 
AND Ubuntu? :-P
```

Original comment by `edgar.b....@gmail.com` on 16 May 2010 at 4:20

**Comment by GoogleCodeExporter** [comment]
```
OK..woops! Scratch that part in my prev comment about turning on Sound Effects 
again 
after launching Pino... I just did that now, after Pino stayed alive for 
several 
minutes, and I made a tweet (about this PA bug, of course) and refreshed 
timeline a 
few times. About 1.5 to 2 seconds after I re-enable Sound Effects, Pino crashed 
with 
the same old 
"Assertion 'pthread_setspecific(t->key, userdata) == 0' failed at 
pulsecore/thread-
posix.c:196, function pa_tls_set(). Aborting.
Aborted"

Oh, well, like I said earlier - I'd rather have Pino working than Sound Effects 
on. 
Sighh..
```

Original comment by `edgar.b....@gmail.com` on 16 May 2010 at 4:28

**Comment by GoogleCodeExporter** [comment]
```
Hope this is solved soon. I had to use mitter because pino is practically 
unusable in
Fedora 12 since 2.6 version.
```

Original comment by `darksh...@gmail.com` on 22 May 2010 at 6:43

**Comment by GoogleCodeExporter** [comment]
```
Removing all calls to Parser.cleanup() does fix the issue. I am however not 
sure at
all what side effects this can have.

See http://0pointer.de/blog/projects/beware-of-xmlCleanupParser for more details

I attach a patch which simply removes these calls.

Cheers,
Julien
```

Original comment by `Valro...@gmail.com` on 23 May 2010 at 12:40

Attachments:
- [002_fix-parser-cleanup.diff](https://storage.googleapis.com/google-code-attachments/pino-twitter/issue-245/comment-12/002_fix-parser-cleanup.diff)

**Comment by GoogleCodeExporter** [comment]
```
I'm testin the patch by ValroffJ. Seems to work OK. So far, no crashes today.
```

Original comment by `darksh...@gmail.com` on 25 May 2010 at 8:44

**Comment by GoogleCodeExporter** [comment]
```
The whole day and no crashes. I had say ValroffJ's patch works. I'll keep 
testing. I 
had sent new and patched RPM packages for Fedora 12 to troorl.

Any Fedora 12 user interested into testing this patch, please try 
http://www.alcancelibre.org/al/aldos/1.2/i686/RPMS.al/pino-0.2.9-2.fc12.al.i686.
rpm
```

Original comment by `darksh...@gmail.com` on 26 May 2010 at 2:24

**Comment by GoogleCodeExporter** [comment]
```
Closed by 7774350dc5 revision (I hope).

Thanks to ValroffJ. So many niceties in the GLib-oriented programming.
```

Original comment by `tro...@gmail.com` on 26 May 2010 at 10:30
- Changed state: **Fixed**

**Comment by GoogleCodeExporter** [comment]
```
On Sabayon too

** (pino:7732): WARNING **: Invalid GValue unboxing (wrong type or NULL)

** (pino:7732): WARNING **: Invalid GValue unboxing (wrong type or NULL)

** (pino:7732): WARNING **: Invalid GValue unboxing (wrong type or NULL)

** (pino:7732): WARNING **: Invalid GValue unboxing (wrong type or NULL)
Assertion 'pthread_setspecific(t->key, userdata) == 0' failed at
pulsecore/thread-posix.c:196, function pa_tls_set(). Aborting.

```

Original comment by `Neal.Qui...@googlemail.com` on 30 May 2010 at 9:07
